src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+        mock_transport
+            .expect_do_send()
+            .returning(|_request_message| {
+                Err(UStatus::fail_with_code(
+                    UCode::UNAVAILABLE,
+                    "transport not available",
+                ))
+            });
+        let client = InMemoryRpcClient::new(Arc::new(mock_transport), new_uri_provider())
+            .await
+            .unwrap();
+
+        // WHEN invoking a remote service operation
+        let message_id = UUID::build();
+        let mut call_options = CallOptions::default();
+        call_options.with_message_id(message_id.clone());
+        let response = client
+            .invoke_method(service_method_uri(), call_options, None)
+            .await;
+
+        // THEN the invocation fails with the error caused at the Transport Layer
+        assert!(response.is_err_and(|e| matches!(e, ServiceInvocationError::Unavailable(_msg))));
+        assert!(!client.contains_pending_request(&message_id));
+    }
+
+    #[tokio::test]
+    async fn test_invoke_method_succeeds() {
+        let message_id = UUID::build();
+        let traceparent = "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01";
+        let mut call_options = CallOptions::default();
+        call_options
+            .with_message_id(message_id.clone())
+            .with_priority(crate::UPriority::UPRIORITY_CS6)
+            .with_ttl(5_000)
+            .with_token("my_token")
+            .with_traceparent(traceparent);
+
+        // GIVEN an RPC client
+        let listener_ref: Arc<Mutex<Option<Arc<dyn UListener>>>> = Arc::new(Mutex::new(None));
+        let response_listener = listener_ref.clone();
+
+        let mut mock_transport = MockTransport::default();
+        mock_transport.expect_do_register_listener().returning(
+            move |_source_filter, _sink_filter, listener| {
+                let mut v = listener_ref.lock().unwrap();
+                *v = Some(listener);
+                Ok(())
+            },
+        );
+        let expected_message_id = message_id.clone();
+        mock_transport
+            .expect_do_send()
+            .withf(move |request_message| {
+                request_message
+                    .attributes
+                    .as_ref()
+                    .map_or(false, |attribs| {
+                        attribs.id.as_ref() == Some(&expected_message_id)
+                            && attribs.priority.value() == UPriority::UPRIORITY_CS6.value()
+                            && attribs.ttl == Some(5_000)
+                            && attribs.token == Some("my_token".to_string())
+                            && attribs.traceparent == Some(traceparent.to_string())
+                    })
+            })
+            .returning(move |request_message| {
+                let request_payload: StringValue = request_message.extract_protobuf().unwrap();
+                let response_payload = StringValue {
```

Comment:
Neat, didn't know about StringValue. That's useful, I can probably clean up some example code :slightly_smiling_face: